File: 0460-lfu-cache/0460-lfu-cache.java
class LFUCache {
    final int capacity;
    int minFreq, curSize;
    Map<Integer, Node> mp;
    Map<Integer, DLL> frequencyMap;

    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.curSize = 0;
        this.minFreq = 0;
        this.mp = new HashMap<>();
        this.frequencyMap = new HashMap<>();
    }
    
    public int get(int key) {
        Node curNode = mp.get(key);
        if (curNode == null) return -1;
        updateNode(curNode);
        return curNode.val;
    }
    
    public void put(int key, int value) {
        if (capacity == 0) return;
        if (mp.containsKey(key)) {
            Node cur = mp.get(key);
            cur.val = value;
            updateNode(cur);
        } else {
            curSize++;
            if (curSize > capacity) {
                DLL list = frequencyMap.get(minFreq);
                mp.remove(list.tail.prev.key);
                list.remove(list.tail.prev);
                curSize--;
            }
            minFreq = 1;
            Node newNode = new Node(key, value);
            DLL newList = frequencyMap.getOrDefault(1, new DLL());
            newList.add(newNode);
            frequencyMap.put(1, newList);
            mp.put(key, newNode);
        }
    }
    
    public void updateNode(Node curNode) {
        int freq = curNode.frequency;
        DLL tmp = frequencyMap.get(freq);
        tmp.remove(curNode);
        if (freq == minFreq && tmp.listSize == 0) {
            minFreq++;
        }
        curNode.frequency++;
        DLL newList = frequencyMap.getOrDefault(curNode.frequency, new DLL());
        newList.add(curNode);
        frequencyMap.put(curNode.frequency, newList);
    }
    
    class Node {
        int key;
        int val;
        int frequency;
        Node next, prev;
        Node (int key, int val) {
            this.key = key;
            this.val = val;
            this.frequency = 1;
        }
    }

    class DLL {
        int listSize;
        Node head, tail;

        DLL() {
            this.listSize = 0;
            this.head = new Node(0, 0);
            this.tail = new Node(0, 0);
            head.next = tail;
            tail.prev = head;
        }

        void add(Node node) {
            Node headNext = head.next;
            node.next = headNext;
            node.prev = head;
            head.next = node;
            headNext.prev = node;
            listSize++;
        }

        void remove(Node node) {
            Node nodePrev = node.prev;
            Node nodeNext = node.next;
            nodePrev.next = nodeNext;
            nodeNext.prev = nodePrev;
            listSize--;
        }
    }

}

File: 0460-lfu-cache/0460-lfu-cache.cpp
class Node {
public:
    int key;
    int val;
    int freq;
    Node* prev;
    Node* next;
    
    Node(int key, int val) {
        this->key = key;
        this->val = val;
        this->freq = 1;
        this->prev = nullptr;
        this->next = nullptr;
    }
};

class DLL {
public:
    Node* head;
    Node* tail;
    int length;
    
    DLL() {
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
        this->length = 0;
    }
    
    void add(Node* node) {
        Node* head_next = head->next;
        head->next = node;
        node->prev = head;
        node->next = head_next;
        head_next->prev = node;
        length++;
    }
    
    void remove(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
        length--;
    }
};

class LFUCache {
public:
    unordered_map<int, DLL*> fwm;
    unordered_map<int, Node*> xt;
    int min_freq, nodes, capacity;
    
    LFUCache(int capacity) {
        this->capacity = capacity;
        this->nodes = 0;
        this->min_freq = 1;
    }
    
    int get(int key) {
        if (xt.find(key) == xt.end()) {
            return -1;
        }
        Node* cur = xt[key];
        update(cur);
        return cur->val;
    }
    
    void put(int key, int value) {
        if (xt.find(key) != xt.end()) {
            Node* cur = xt[key];
            cur->val = value;
            update(cur);
        } else {
            DLL* list = nullptr;
            nodes++;
            if (nodes > capacity) {
                list = fwm[min_freq];
                xt.erase(list->tail->prev->key);
                list->remove(list->tail->prev);
                nodes--;
            }
            min_freq = 1;
            Node* new_node = new Node(key, value);
            if (fwm.find(1) == fwm.end()) {
                list = new DLL();
            } else {
                list = fwm[1];
            }
            
            list->add(new_node);
            fwm[1] = list;
            xt[key] = new_node;
        }
    }
    
    void update(Node* node) {
        int frequency = node->freq;
        DLL* list = fwm[frequency];
        list->remove(node);
        fwm[frequency] = list;
        
        if (frequency == min_freq && list->length == 0) {
            min_freq++;
        }
        
        node->freq++;
        if (fwm.find(node->freq) == fwm.end()) {
            list = new DLL();
        } else {
            list = fwm[node->freq];
        }
        
        list->add(node);
        fwm[node->freq] = list;
    }
};